framework/src/main/java/org/checkerframework/common/wholeprograminference/SceneToStubWriter.java

```diff
@@ -0,0 +1,605 @@
+        this.basetypes = basetypes;
+        this.types = types;
+        this.enumConstants = enumConstants;
+    }
+
+    /**
+     * Writes the annotations in {@code scene} to {@code out} in stub file format.
+     *
+     * @param scene the scene to write out
+     * @param basetypes a map from the {@code description} field of {@code ATypeElement}s to the
+     *     {@code TypeMirror}s that represent their base Java types
+     * @param types a map from fully-qualified names to the {@code ATypeElement}s representing their
+     *     declarations
+     * @param enumConstants a map from fully-qualified enum names to the enum constants defined in
+     *     that name
+     * @param out the Writer to output the result to
+     */
+    public static void write(
+            AScene scene,
+            Map<String, TypeMirror> basetypes,
+            Map<@FullyQualifiedName String, TypeElement> types,
+            Map<@FullyQualifiedName String, List<VariableElement>> enumConstants,
+            Writer out) {
+        SceneToStubWriter writer = new SceneToStubWriter(basetypes, types, enumConstants);
+        writer.writeImpl(scene, new PrintWriter(out));
+    }
+
+    /**
+     * Writes the annotations in {@code scene} to the file {@code filename} in stub file format.
+     *
+     * @param scene the scene to write out
+     * @param basetypes a map from the {@code description} field of {@code ATypeElement}s to the
+     *     {@code TypeMirror}s that represent their base Java types
+     * @param types a map from fully-qualified names to the {@code ATypeElement}s representing their
+     *     declarations
+     * @param enumNamesToEnumConstant a map from fully-qualified enum names to the enum constants
+     *     defined in that name
+     * @param filename the path of the file to write to
+     * @throws IOException if the file doesn't exist
+     * @see #write(AScene, Map, Map, Map, Writer)
+     */
+    public static void write(
+            AScene scene,
+            Map<String, TypeMirror> basetypes,
+            Map<@FullyQualifiedName String, TypeElement> types,
+            Map<@FullyQualifiedName String, List<VariableElement>> enumNamesToEnumConstant,
+            String filename)
+            throws IOException {
+        write(scene, basetypes, types, enumNamesToEnumConstant, new FileWriter(filename));
+    }
+
+    /**
+     * Returns the part of a binary name that specifies the package.
+     *
+     * @param className the binary name of a class
+     * @return the part of the name referring to the package
+     */
+    private static String packagePart(@BinaryName String className) {
+        int lastdot = className.lastIndexOf('.');
+        return (lastdot == -1) ? "" : className.substring(0, lastdot);
+    }
+
+    /**
+     * Returns the part of a binary name that specifies the basename of the class. This method
+     * replaces the {@code $}s in the names of inner classes with {@code .}s, so that they can be
```

Comment:
I don't think this method should do the replacement.  In fact, the replacement should never be necessary since you parse apart the result anyway.  It is less error-prone to work with binary names rather than converting to fully-qualified names.  That change will make the first sentence of the documentation accurate, as well.